HSGP: Birthdays Example

juanitorduz opened the pull request:
Closes https://github.com/pymc-devs/pymc-examples/issues/626

**Scope:**
- A simpler and detailed version of the model presented in https://juanitorduz.github.io/birthdays/ (@juanitorduz )
- The complete model at the end (@bwengals ). Bill, you can use the code from my post to generate the features of the special dates.

TODO:
- [x] Intro
- [x] EDA
- [x] Simple model
- [ ] Complete model
- [x] References
- [x] Additional packages in metadata. 

<!-- readthedocs-preview pymc-examples start -->
----
📚 Documentation preview 📚: https://pymc-examples--627.org.readthedocs.build/en/627/

<!-- readthedocs-preview pymc-examples end -->

juanitorduz commented:
The simple model is ready for a first round of review.

juanitorduz commented:
Well, we decided with Bill that we will present a simpler model to illustrate HSGP and then Bill will add at the end the complete model from Aki. I was not able to add the Horseshoe prior (very bad r hats and divergences) in my post so Bill will give it a go :)

All feedback is welcome!

AlexAndorra commented:
Ok, so it sounds like it's your post updated + simpler model, so I'll definitely review this PR instead! Please feel free to ping me when it's ready for review

juanitorduz commented:
As discussed with Bill, we want to keep the model specification closer to the Stsan one for this example.  In https://github.com/pymc-devs/pymc-examples/pull/627/commits/08f4bc8ac19b700206c89737dc8fa27db81f0cb4 , I changed the `day_of_week` parametrization from zero-sum-normal to one-hot-encoding setting Monday coefficient to zero. Note this corresponds to [`Model 3: Slow trend + yearly seasonal trend + day of week`](https://avehtari.github.io/casestudies/Birthdays/birthdays.html#Model_3:_Slow_trend_+_yearly_seasonal_trend_+_day_of_week). In particular, if you go to the Stan code we see there is actually not an intercept as we are modeling the standardized births (`vector[N] intercept = 0.0 + f_day_of_week[day_of_week]`, see https://github.com/avehtari/casestudies/blob/master/Birthdays/gpbf3.stan#L47C3-L47C58)

juanitorduz commented:
In view of https://github.com/pymc-devs/pymc-examples/issues/626#issuecomment-1911777083 I think the very first iteration is ready for review in order to collect feedback.

juanitorduz commented:
The pre-commit-ci jobs fails with a strange error. Still the pre-commit on the GitHub actions passes :)

bwengals commented:
Left a bunch of comments, pls keep in mind that I didnt think about what should be changed now vs. another iteration, so they're all pretty optional.

One overall though, is it doesn't really follow the Bayesian workflow.  It starts 1. EDA, 2. Model, 3. Results.  But to me the step from the EDA straight to (a pretty complicated) model wouldn't have followed.  Would it be possible to interleave EDA, building the model in steps, looking at results?  Right now it feels like we're just copying the Vehtari case study model 3 but at the same time pretending we arrive at it naturally?  Another option would be to just explicitly say we are copying model 3 from Vehtari, and then not do all the EDA.

theorashid commented:
I would vote for copying straight from Vehtari à la numpyro example, skipping the EDA. My attention span isn't long enough for a really long example – I'm just here to copy and paste the HSGP code. Cite the original blog post if they want the workflow

AlexAndorra commented:
Oooh fantastic, thanks Juan, I'll review that ASAP -- probably this
weekend, if you can wait.

Strongly disagree with just copy pasting the code, especially for such
involved models, where EDA and contextualization help make sense of the
tradeoffs and modeling choices.
Not to mention that having to switch from one post to another to get
essential information disturbs flow state and decreases attention span even
more

El El mar, 30 ene 2024 a la(s) 11:21, theorashid ***@***.***>
escribió:

> I would vote for copying straight from Vehtari à la numpyro example,
> skipping the EDA. My attention span isn't long enough for a really long
> example – I'm just here to copy and paste the HSGP code. Cite the original
> blog post if they want the workflow
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pymc-devs/pymc-examples/pull/627#issuecomment-1916974594>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHIJMTAMBFYOIAHORE2SIR3YRD6XTAVCNFSM6AAAAABB47QFPCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSMJWHE3TINJZGQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

AlexAndorra commented:
<p>Agree with Bill</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915414954'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<p><s>patters</s> <strong>patterns</strong></p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415092'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<p>I think it's just a ReviewNB limitation</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415219'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra reviewed (changes requested):
I really love this example, well done @juanitorduz 👏 Can't wait to review the complete, complex model!
I left a bunch of comments -- LMK if anything is unclear.
In general, I agree with all of @bwengals 's comments in ReviewNB, as well as his more general point about the workflow and iterative model-building

juanitorduz commented:
Thank you all for the detailed review! I will address them at the end of the week 💪 . Much appreciated :)

juanitorduz commented:
<p>I am not 100%, maybe to have smaller numbers (maybe easy to think about priors?)</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915414977'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>Absolutely, I willl tackle this at the end after we have a better picture of the scope of this first iteration.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915414954'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>yes! added!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915414996'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>Good point!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415037'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>Indeed! This is a key observation from Aki's blog:</p><p> </p><blockquote>Looking at the time series of whole data we see the dots representing the daily values forming three branches that are getting further away from each other. In previous analysis (BDA3) we also had a model component allowing gradually changing effect for day of week and did observe that the effect of Saturday and Sunday did get stronger in time</blockquote><p> </p><p>which motivates the amplitude GP. As we wont add this component for now... shall I remove this plot?</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415067'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>Added a commet:</p><p>"We want to work on the normalized log scale of the relative births. The reason for this is to work on a scale where is easier to set up priors (scaled space) and so that the heteroscedasticity is reduced (log transform)."</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415118'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>This is also done by Aki.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415118'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>Thanks for the feedback! Here are some remarks:</p><p> </p><ul><li>Build the model iteratively: Ideally yes, but then this notebook would be super long and we would be replikating Aki's complete blog post. My suggestion (intention) is to make the EDA clear enough so thatstarting with this baseline model (Model 3 from Aki's blog) is not unreasonable even for a beginner. What do you think?</li><li>Dummy encoding + intercept: You are right! this is a comment which got mixed with the initial ZSN approach where I intentionally removed an intercept.</li><li>Regarding the ZSN parametrization: I also like it better but might be harder to interpret for new users. I suggest we keep the one-hot encoding to make it similar to Aki's approach and to keep the new topics (for new users) bounded. We can add the ZSN approach as a comment. Also, note we do not have even a beginners example to ZSM. We should add this into the to-to list ;)</li></ul>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415164'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>Thanks!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1928064899'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>Yes, It does appear on the docs <a href="https://pymcio--627.org.readthedocs.build/projects/examples/en/627/gaussian_processes/GP-Births.html" target="_blank">https://pymcio--627.org.readthedocs.build/projects/examples/en/627/gaussian_processes/GP-Births.html</a></p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415219'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>Done!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415261'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>added a clarification</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415279'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>Added!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415313'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>An argument for the one-hot encoding is because we can replicate Aki's results (see comment  just after the trace).</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1928064943'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>Thanks! added!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415337'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>You are right about this looking a bit weird. Once I got it it became my default as I love sklearn transformers. I will try to simplify ...</p><p> </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415359'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>You are right! I changed it. (somehow I wanted to remain optimistic). I added :</p><p> </p><p>This does not seem very good as there is a pretty big discrepancy between black line and shaded blue in the bulk of posterior, tails look good. This suggests we might be missing some covariates. We explore this in a latter more complex model.</p><p> </p><p>Which should be a nice entry point for the next iteration where you work out the complete model.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415376'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>solved! thanks!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1928064964'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>yes! based on the comments above a function will help so that people don't get lost in plotting code.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1928064979'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<p>We could try and see how modeling on the raw outcome scale works?</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915414977'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<p>I think for the weekend it's easy to imagine: less staff, some births are postponed to the beginning of the following week. This dataset is quite old, but I'm guessing this pattern is even more apparent in recent years, with the increasing technological ability to trigger labor</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415048'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<p>I think so, otherwise the EDA is overwhelming. We can add back this plot back when / if we add that component -- that'd be useful context</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415067'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<blockquote>make the EDA clear enough so thatstarting with this baseline model (Model 3 from Aki's blog) is not unreasonable even for a beginner.</blockquote><p>SGTM</p><p> </p><blockquote>We can add the ZSN approach as a comment. </blockquote><p>Yes, I think we can show the equivalent parametrization with ZSN in a markdown cell. It's a good way for people to start getting familiar with ZSN.</p><p> </p><blockquote>Also, note we do not have even a beginners example to ZSM. We should add this into the to-to list ;)</blockquote><p>I know I know, I've been meaning to do that for months now, but have been able to set the time aside yet -- shame on me ;)</p><p> </p><p>On the workflow more generally, what do you think of adding the corresponding modeling component after each corresponding EDA plot? That way, it's really not a surprise to readers at that point, and this part here is just a summary of what we already saw</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415164'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<p>Yep, replied above, I think we have a good solution now</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1928064943'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<p>See our private conversation on that point</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1928064919'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<p>I didn't know about that! Really good to know.</p><p>You're not using <code>drop_first</code> above though </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415238'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<p>Thanks Bill! Lots of interesting threads here. To make sure we're on the same page, we're talking about these 2 different parametrizations, right?</p><p> </p><ul><li><strong>Dummy encoding</strong></li></ul><p> </p><pre class="ql-syntax" spellcheck="false"># day of week
b_day_of_week_no_monday = pm.Normal(
&nbsp;&nbsp;&nbsp;&nbsp;name="b_day_of_week_no_monday", sigma=1, dims="day_of_week_no_monday"
)
b_day_of_week = pt.concatenate(([0], b_day_of_week_no_monday))

# then used in the linear predictor as:
b_day_of_week[day_of_week_idx_data] * (day_of_week_idx_data &gt; 0)
</pre><p> </p><ul><li><strong>ZSN</strong> (taken from <a href="https://juanitorduz.github.io/birthdays/" target="_blank">Juan's blogpost</a>)</li></ul><p> </p><pre class="ql-syntax" spellcheck="false">log_f_day_of_week = gp_day_of_week.prior(
    name="log_f_day_of_week", X=normalized_obs_data[:, None], dims="obs"
) # why is it log BTW?
f_day_of_week = pm.Deterministic(
    name="f_day_of_week", var=pt.exp(log_f_day_of_week), dims="obs"
)

b_day_of_week = pm.ZeroSumNormal(name="b_day_of_week", sigma=1, dims="day_of_week")

# then used in the linear predictor as:
f_day_of_week * b_day_of_week[day_of_week_idx_data]
</pre><p> </p><p>First, there are <strong>2 things I don't get here</strong>: 1) Why does this one has a GP and the previous one doesn't? 2) Why is the GP exponentiated? If somebody knows, any help is appreciated :)</p><p> </p><p>That put aside, <strong>these parametrizations are indeed equivalent</strong>: the first one is using pivoting (fixing one category to 0), while the second one is using ZSN (fixing the sum of the categories to 0). </p><p>It changes the interpretation and the code, but both params are used to prevent over-parametrization of the model, because we have an intercept (well, just a GP for the global trend here, but that's the same idea).</p><p>So, to answer your second question Bill: there is <strong>no partial pooling</strong> with ZSN here; it's just removing one degree of freedom in a different way than pivoting (aka dummy encoding).</p><p> </p><p>I do agree with you that day of the week is a nice use case for ZSN, because you're never going to predict for some new 8th day of the week, but I do think this point is more appropriate for a ZSN-focused example.</p><p> </p><p>All this makes me wonder though: <strong>why don't we one-hot encode day of week and pass that to the GP?</strong> That would mean <code>input_dim = X.shape[1] = 1 + 7 = 8</code>. Would that work? Would it be equivalent to the previous parametrization?</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415164'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
Thank you for all the feedback! I will look into the upcoming days! This PR will be my priority next week! 💪

bwengals commented:
<blockquote>First, there are&nbsp;<strong>2 things I don't get here</strong>: 1) Why does this one has a GP and the previous one doesn't? 2) Why is the GP exponentiated? If somebody knows, any help is appreciated :)</blockquote><p>I think it's because in the original case study here <a href="https://avehtari.github.io/casestudies/Birthdays/birthdays.html" target="_blank">https://avehtari.github.io/casestudies/Birthdays/birthdays.html</a>, Model 3 doesnt include a time component on day of week (day of week effects are treated as constant over the years), Model 4 introduces a time component based on the hypothesis that they day of week effect grows more pronounced as time goes on.</p><p> </p><blockquote><strong>these parametrizations are indeed equivalent</strong></blockquote><p>OK... I had to think about this a while, I think I've been super confused about ZSN on this point!  If you pass a prior for sigma for ZSN, it represents a random effect with a zero sum constraint and with partial pooling.  If you don't, and pass a specific value for sigma, it represents a fixed effect with a zero sum constraint and no partial pooling, like you said.  Still have more ZSN Q's... but it's off topic.</p><p> </p><blockquote>All this makes me wonder though:&nbsp;<strong>why don't we one-hot encode day of week and pass that to the GP?</strong>&nbsp;That would mean&nbsp;<code>input_dim = X.shape[1] = 1 + 7 = 8</code>. Would that work? Would it be equivalent to the previous parametrization?</blockquote><p>I think you could.  Is that extra dimension, the 1 in 1 + 7 = 8, is that time?  That would give you the interaction of day of week and time.  The size of the data might constrain you here, and also the hypothesis in described in Vehtari's model 4 is a bit more structured.  But, yah I think it'd be reasonable thing to do.</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415164'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<blockquote>Model 3 doesnt include a time component on day of week (day of week effects are treated as constant over the years), Model 4 introduces a time component based on the hypothesis that they day of week effect grows more pronounced as time goes on.</blockquote><p>Ah ok, makes sense, thanks Bill!</p><p> </p><blockquote>If you pass a prior for sigma for ZSN, it represents a random effect with a zero sum constraint and with partial pooling. If you don't, and pass a specific value for sigma, it represents a fixed effect with a zero sum constraint and no partial pooling, like you said. Still have more ZSN Q's...</blockquote><p>Yes, exactly! Feel free to send them my way in DM ;)</p><p> </p><blockquote>Is that extra dimension, the 1 in 1 + 7 = 8, is that time? </blockquote><p>Yes</p><p> </p><blockquote>That would give you the interaction of day of week and time. The size of the data might constrain you here [...] But, yah I think it'd be reasonable thing to do.</blockquote><p>Yeah and HSGP would probably break with 8 input dimensions. But good to know it'd be reasonable!</p><p>In general, do you recommend more this "big GP" parametrization, or the ZSN + GP parametrization from Juan's blogpost?</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415164'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>ok! I will remove it!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415067'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>This is such en enlighting threat! Thanks! In particular, Alex</p><blockquote>If you pass a prior for sigma for ZSN, it represents a random effect with a zero sum constraint and with partial pooling. If you don't, and pass a specific value for sigma, it represents a fixed effect with a zero sum constraint and no partial pooling, like you said.&nbsp;</blockquote><p>I did not know this! This deserves a proper explanation in your upcoming ZSN notebook ;)</p><p> </p><p>Regarding the scope of this example for this iteration. My suggestion is let's keep it simple because we do now we want to iterate over this example as described in the corresponding issue. I won't close this thread so that we have it for reference. I suggest keeping the one-hot encoding to reproduce Aki's results of model 3 (therefore I will kipp the ./100 scaling). I will add a comment on the "equivalent"  ZSN parametrization as a FYI and link to my post if that's ok with you guys :)</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415164'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>thanks!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1957352382'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>Agree! I hope its better now! </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415238'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>To keep the scope of the notebook: Reproduce Model 3 from Aki's blog (see discussion below regarding modeling choices), I suggest we keep the scaling ;)</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915414977'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
@bwengals @AlexAndorra Thank you very much for all the valuable feedback! I think I have addressed most of the comments. In particular, I simplified the code, re-named `obs` -> `time` variables, and did a complete re-write of the introduction in view of the scope of this notebook as part of a bigger sequence of notebooks regarding this example (see the associated issue). I am ready for another review round. I think the example is in much better shape than the initial versions. 

Once we are done with the feedback, I will trim the annoying matplotlib output of the plots (`Text( ...`)), since this is just my auto-fomatter. I'll fix it at the end, I promise 😅 .

AlexAndorra commented:
<p>Sounds good. Then maybe mention that it's done to be as close as possible to Aki's case study, so that people understand they don't <em>have</em> to scaled the data this way for HSGP to work</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915414977'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<p>Typo is still there</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415092'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

AlexAndorra commented:
<ul><li>Typo: "<s>whe</s> <strong>we</strong> want to consider"</li><li>And I think the log in the mu of the LogNormal is because you're setting the prior on the normalized log scale, because that's what the GP is seeing, not because of the LogNormal</li></ul>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415192'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
> <ul><li>Typo: "<s>whe</s> <strong>we</strong> want to consider"</li><li>And I think the log in the mu of the LogNormal is because you're setting the prior on the normalized log scale, because that's what the GP is seeing, not because of the LogNormal</li></ul>
> 
> ---
> 
>  <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415192'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

Yes! This is what I wanted to say 🙈. I'll clarify it better thanks !

AlexAndorra reviewed (changes requested):
Gave another round of review @juanitorduz -- looks really awesome 😍 
Just a last round of changes and then I think we're in a good position to merge

juanitorduz commented:
<p>Good point! MMM I think I then will use a trace plot as I don't want to introduce more noew concepts ... (people might get distracted)</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1988463017'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>👍 👌</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415192'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>Good point!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1988462938'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>😄. ok, not it is fixed!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415092'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>👍</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1988462916'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>ok! 👍</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915414977'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>ok thanks!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1988462849'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

bwengals commented:
<p>maybe missed it earlier, but I do think it'd be nice to say something like "Next we're going to systematically look at seasonality, first yearly, then monthly" ... etc.  Just wanted to bring it up again in case it got lost in the shuffle -- also totally ok if you think it messes your flow</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415037'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

bwengals commented:
<p>Wanna re-mention this one:</p><p> </p><blockquote>All of these building blocks should not come as a surprise after looking into the EDA section.</blockquote><p> </p><p>I think it's better, especially for beginners, to not say things are obvious or not a surprise, because it doesn't help understanding and only potentially makes them feel dumb. To be honest after doing the EDA above, I dunno if I'd make the exact same model (unless I was copying the Vehtari case study).</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415164'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>No, it is a box-cell, see the rendered docs <a href="https://pymcio--627.org.readthedocs.build/projects/examples/en/627/gaussian_processes/GP-Births.html" target="_blank">https://pymcio--627.org.readthedocs.build/projects/examples/en/627/gaussian_processes/GP-Births.html</a> ;)</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-2024048779'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>You are 100% right! Let me rephrase it 🙏</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415164'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>Let me know what you think about the change :)</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-1915415164'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>done!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-2026323177'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>Added! Let me know what you think :)</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-2026323184'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>Done!</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-2024048724'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>I also added a link to theis note about the source of the birthdays dataset <a href="https://statmodeling.stat.columbia.edu/2020/10/25/birthday-data" target="_blank">https://statmodeling.stat.columbia.edu/2020/10/25/birthday-data/</a></p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/627/discussion/#comment-2024048724'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
thank you for your comments @bwengals ! I think I addressed them all. Let me know whet you think about the changes :)

bwengals commented:
<p>One last detail in the conclusion:  The complexity for HSGP is $O(nm + m)$ instead of $O(nm^2)$.  From the "Practical HSGP" paper on efficiency:</p><blockquote><br></blockquote><blockquote>&nbsp;While the pre-computation cost of the basis functions is O(m2n), the compu-</blockquote><blockquote>tational cost of learning the covariance function parameters is</blockquote><blockquote>O(mn + m) in every step of the optimizer or sampler. This is</blockquote><blockquote>a big advantage in terms of speed for iterative algorithms such</blockquote><blockquote>as Markov chain Monte Carlo (MCMC). Another advantage</blockquote><blockquote>is the reduced memory requirements of automatic differentia-</blockquote><blockquote>tion methods used in modern probabilistic programming frame-</blockquote><blockquote>works, such as Stan (Carpenter et al., 2017) and others. This is</blockquote><blockquote>because the memory requirements of automatic differentiation</blockquote><blockquote>scale with the size of the autodiff expression tree which in direct</blockquote><blockquote>implementations is simpler for basis function than covariance</blockquote><blockquote>matrix-based approaches</blockquote>

juanitorduz commented:
Thank you @bwengals ! Fixed in https://github.com/pymc-devs/pymc-examples/pull/627/commits/adf217073cf427b7e2a4aa0a168ea6ade85b5aa6 :)

AlexAndorra commented:
Shall I merge @juanitorduz ?

juanitorduz commented:
> Shall I merge @juanitorduz ?

Yes :) 🙏
